src/Microsoft.Identity.Client/ITokenCache.cs

```diff
@@ -49,10 +52,43 @@ public interface ITokenCache
         /// Sets a delegate called before any library method writes to the cache. This gives an option to the delegate
         /// to reload the cache state from a row in database and lock that row. That database row can then be unlocked in the delegate
         /// registered with <see cref="SetAfterAccess(TokenCacheCallback)"/>
+        /// If you need async/task-based callbacks, please use SetBeforeWriteAsync instead.
         /// </summary>
         /// <param name="beforeWrite">Delegate set in order to prepare the cache serialization</param>
         void SetBeforeWrite(TokenCacheCallback beforeWrite);
 
+        /// <summary>
+        /// Sets a delegate to be notified before any library method accesses the cache. This gives an option to the
+        /// delegate to deserialize a cache entry for the application and accounts specified in the <see cref="TokenCacheNotificationArgs"/>.
+        /// See https://aka.ms/msal-net-token-cache-serialization.
+        /// This provides the same functionality as SetBeforeAccess but it provides for an async/task-based callback.
+        /// </summary>
+        /// <param name="beforeAccess">Delegate set in order to handle the cache deserialization</param>
+        /// <remarks>In the case where the delegate is used to deserialize the cache, it might
+        /// want to call <see cref="DeserializeMsalV3(byte[], bool)"/></remarks>
+        void SetBeforeAccessAsync(Func<TokenCacheNotificationArgs, Task> beforeAccess);
```

Comment:
I'm in 2 minds here. The Async prefix is usually for Async methods (here this is the delegate which is async. But SetAsyncBeforeAccess is probably not better?

Reply:
I have the same existential issue.  I've renamed them to the XXXAsync methods since henrik's comment on discoverability (SetBefore... will pop up two methods in intellisense, one SetBeforeAccess and another SetBeforeAccessAsync to show that there are two versions).